### ISSUE TITLE
Change from just "Loading..." to an actual spinner

### ISSUE THREAD

**Issue by cynthiateeters** [novelrt/novelrt.github.io#52]
The time delay in getting data back from our custom GitHub contributor API is such that it needs a visual notification to the user.

There is a npm library [React Spinners](https://github.com/davidhu2000/react-spinners) that offers a choice of spinners.

We could use their Pacman spinner with a cute saying as is seen in the below gif.

I respectfully request that I be assigned to this issue.

![pacman](https://user-images.githubusercontent.com/13385801/170846664-dab9bc42-5985-45eb-ae0f-7b03ecdf87d3.gif)

**Comment by RubyNova** [comment]
Sounds good to me! Assigned. :D

**Comment by capnkenny** [comment]
10/10 on the Pac-man spinner and sayings

**Comment by cynthiateeters** [comment]
I have 2 observations:

1. The contributors data doesn't have large changes. (Yes, commit numbers will change and, from time to time, a new person with a small number of commits will appear close to the bottom of the list.)
2. The list on the website could be *slightly out-of-sync* with the current data without the world come crashing down.

This suggests that we should take the load of making the API call off of the Gatsby page. Instead, put the effort of calling the API somewhere else and store the resulting data in a JSON file. 

The Gatsby page could then do a fetch from that JSON file and its response time to the user will be much faster.

I have successfully written a GitHub action in my fork that schedules a call to the API and saves the results into a file called contributors.json in the root of the repo. 

Then I modified the fetch inside the Gatsby page to load that file. 

**It is indeed much faster.**

By compromising on the frequency of the scheduled GitHub action, it will also have the benefit of limiting calls to your Azure account.

**Comment by capnkenny** [comment]
Based on this, we could do something similar to what we have currently for Doxygen, where we can:
- Make a new branch
- Update the contributors.json file
- Auto merge the change based on a rule we set within the repo

Considering it'd probably be as simple as using a cURL request to grab the most up to date json, this wouldn't be terribly hard to set up.

@RubyNova any issues with this approach? The only ones I see are maintenance (which is minimal) and its not "real time" (but we don't have PRs pumping into our repos every hour so I think it's a non-issue)

**Comment by cynthiateeters** [comment]
I agree with using curl. My GH Action doesn't create a new branch, but its logic is similar to what @capnkenny is suggesting.

```yaml
# This is a basic workflow to help you get started with Actions

name: Fetch Contributors

# Controls when the workflow will run
on:
  # Triggers the workflow on push or pull request events but only for the main branch
  push:
    branches: [ main ]
    
  schedule:
    - cron: "05 3 * * *"

  # Allows you to run this workflow manually from the Actions tab
  workflow_dispatch:

# A workflow run is made up of one or more jobs that can run sequentially or in parallel
jobs:
  # This workflow contains a single job called "fetch"
  fetch:
    # The type of runner that the job will run on
    runs-on: ubuntu-latest

    # Steps represent a sequence of tasks that will be executed as part of the job
    steps:
      # Checks-out your repository under $GITHUB_WORKSPACE, so your job can access it
      - uses: actions/checkout@v3

      # Runs a single command using the runners shell
      - name: Run curl
        run: curl https://novelrt-functions.azurewebsites.net/api/novelrtcontributors -o contributors.json
        
      - name: Set email
        run: git config --global user.email "cynthiateeters@gmail.com"
        
      - name: Set username
        run:  git config --global user.name "GitHub Action"
        
      - name: stage changed files
        run: git add .
      - name: commit changed files
        run: git commit -m "Auto updating contributors.json"
      - name: fetch from main
        run: git fetch origin main
      - name: push code to main
        run: git push origin HEAD:main
```

**Comment by RubyNova** [comment]
@capnkenny what does this mean for the existing azure function?

**Comment by capnkenny** [comment]
@RubyNova We would still use it, it would just be used by the action and only hit so often instead of on every page load.

Not only helps with site performance, but reduces the chance someone spams the function and drives your bill up :^)

**Comment by RubyNova** [comment]
I see.

In that case, I think we should just go ahead and resolve https://github.com/novelrt/NovelRT-Azure-Functions/issues/3 by committing to .NET 6.

The concerns outlined in the ticket are probably no longer even relevant if we do this.

**Comment by capnkenny** [comment]
I'm not entirely buying that switching the function to .NET would alleviate the concern here. Not without at least seeing a comparison with what we have now

**Comment by RubyNova** [comment]
Why wouldn't it?

the function could be written in ALGOL for all the GHA could care for, no?

**Comment by capnkenny** [comment]
> Why wouldn't it?
> 
> 
> 
> the function could be written in ALGOL for all the GHA could care for, no?

Yes, GHA would not care. But are you agreeing to switch to the GHA update? Or are you saying to not resolve this and just update the function?

I may be misunderstanding but it sounds like we're conflating two issues that are somewhat independent of each other.

**Comment by RubyNova** [comment]
They are not independent of each other.

Currently the Azure function takes x amount of time to return the list. This is mostly due to node.js + contacting GH API, probably. To reiterate my concerns of the other ticket:

- I've been told that node.js supposedly has 0 startup time but I don't entirely buy into that. It's Node. Nothing about it is fast.
- .NET has a startup time, but I know that its startup times have been reduced as of .NET 6.
- The majority of the maintainer team doesn't really write JS and this is a back end thing reaching into our org's statistics; this is the most important thing for me here.

If we do this switch to the JSON file with a GHA that just, say, runs nightly or something, then .NET and node startup times are no longer even relevant and we get the benefit of having the Azure function codebase for better accessibility for the maintainer teams. My main line of thinking here is that since the guy who volunteered to help us maintain the website above and beyond being a regular contributor is no longer around, it would basically be all on you to maintain the function as it stands in the current codebase.

That's not fair to you and we can't expect any contributor to just stand up and take that heightened responsibility right now.

Therefore, for the sake of everyone involved, I think the azure function should be rewritten to .NET 6. I'm the one currently handling building and deploying of it anyway. I'm like a lemon when it comes to Node. The GHA doesn't care what it's written in, but other humans somewhat do, given the current situation.

TL;DR: the azure function is currently abandonware. Lol.

**Comment by cynthiateeters** [comment]
If you are looking at making a couple GH API calls once daily, you will not come close to hitting GH's anonymous rate limitations. That means it could be done without any need for GH secrets. We could write a script that the scheduled GHA runs to make the JSON file.

**Comment by RubyNova** [comment]
Then there's a misunderstanding on my part, apologies.

How often were you thinking this JSON file would get updated?

**Comment by cynthiateeters** [comment]
For example: Here's the endpoint for anonymously getting the contributors for the one repo. 

```
https://api.github.com/repos/novelrt/NovelRT/contributors
```

**Comment by cynthiateeters** [comment]
As Kenny is saying, maybe we do once per day. Even once every 6 hours or once per hour should not be rate limiting.

**Comment by cynthiateeters** [comment]
We have a list of repos we want to include and loop through them creating a single JSON file. Then the Gatsby page reads and parses the file.
